mirai-console/frontend/mirai-console-terminal/src/LoggingService.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2019-2022 Mamoe Technologies and contributors.
+ *
+ * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
+ * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
+ *
+ * https://github.com/mamoe/mirai/blob/dev/LICENSE
+ */
+
+package net.mamoe.mirai.console.terminal
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.onFailure
+import net.mamoe.mirai.utils.TestOnly
+import java.io.File
+import java.io.RandomAccessFile
+import java.nio.file.Files
+import java.time.Instant
+import java.time.ZoneId
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.Executors
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.atomic.AtomicReference
+
+internal sealed class LoggingService {
+    @TestOnly
+    internal lateinit var switchLogFileNow: () -> Unit
+
+    internal abstract fun pushLine(line: String)
+}
+
+internal class LoggingServiceNoop : LoggingService() {
+    override fun pushLine(line: String) {
+    }
+
+    init {
+        @OptIn(TestOnly::class)
+        switchLogFileNow = {}
+    }
+}
+
+@OptIn(ConsoleTerminalExperimentalApi::class)
+internal class LoggingServiceI : LoggingService() {
+
+    private val threadPool = Executors.newScheduledThreadPool(3, object : ThreadFactory {
+        private val group = ThreadGroup("mirai console terminal logging")
+        private val counter = AtomicInteger(0)
+        override fun newThread(r: Runnable): Thread {
+            return Thread(
+                group,
+                r,
+                "Mirai Console Terminal Logging Thread#" + counter.getAndIncrement()
+            ).also { thread ->
+                thread.isDaemon = true
+            }
+        }
+    })
+    private val threadDispatcher = threadPool.asCoroutineDispatcher()
+    private val pipeline = Channel<String>(capacity = 2048)
```

Comment:
这将会创建大小为 2048 的 array ---- 很明显多余了

使用 Channel 默认的 BUFFERED 就足够
```suggestion
    private val pipeline = Channel<String>(capacity = Channel.BUFFERED)
```

Reply:
使用大容量是经过考虑的，容量过小会在记录大量日志的时候导致日志文件顺序错乱

Reply:
没看到大量日志的 test

可以试试在 trySend fail 的时候开一个协程, 用 start=UNDISPATCHED, 这样可能可以保证顺序, 不过启动协程/提交任务给线程还是不如缓冲效率高, 缓冲也要有

Reply:
test是在我本地测试的，因为大量日志写CI并不好写（无法对随机性做出预测）

如果是 undispatched 那么尝试提交依然还在原线程，此时无开新携程必要，并且我选择了failed后使用线程池提交是因为并不想堵住原线程（日志提交者），综合考虑下来我选择了最好的大缓冲区，并且占用也不高

Reply:
大缓冲区仍然会有顺序问题, 只有从根本解决

UNDISPATCHED 会在原线程跑到第一个挂起点 (suspension point), 这个挂起点就会是 channel.send, 只是我在猜它能不能像 start=ATOMIC 那样保证协程启动顺序因此保证 send 顺序, 因为 ATOMIC 现在是 experimental.

Reply:
如果使用此策略，那么原执行速度会很大程度被磁盘IO影响, 尤其是这种零碎数据写入, 我没使用能保证顺序的直接写入文件也是基于此原因

Reply:
不会影响, 写操作仍然是单独在那个 loop 里跑. 这只修改 pushInPool

`AbstractChannel.send`:
```kotlin
    public final override suspend fun send(element: E) {
        // fast path -- try offer non-blocking
        if (offerInternal(element) === OFFER_SUCCESS) return  // 直接 return
        // slow-path does suspend or throws exception
        return sendSuspend(element) //
    }

    // 这里必定挂起协程, 必定导致 pushInPool launch 的协程挂起, 然后 pushInPool 会 return.
    private suspend fun sendSuspend(element: E): Unit = suspendCancellableCoroutineReusable sc@ { cont ->
    }
```

Reply:
写个 test 试下就知道了, 总是得有大量日志的 test 的